ecc/goldilocks/decaf.go

```diff
@@ -0,0 +1,135 @@
+package goldilocks
+
+import fp "github.com/cloudflare/circl/math/fp448"
+
+// Decaf provides operations of a prime-order group from goldilocks curve.
+// Its internal implementation uses the twist of the goldilocks curve.
+// This uses version 1.1 of the encoding. Decaf is a zero-length datatype.
+type Decaf struct{ c twistCurve }
+
+// Elt is an element of the Decaf group. It must be always initialized using
+// one of the Decaf functions.
+type Elt struct{ p twistPoint }
+
+func (e Elt) String() string { return e.p.String() }
+
+// IsValid returns True if a is a valid element of the group.
+func (d Decaf) IsValid(a *Elt) bool { return d.c.IsOnCurve(&a.p) }
+
+// Identity returns the identity element of the group.
+func (d Decaf) Identity() *Elt { return &Elt{*d.c.Identity()} }
+
+// Generator returns the generator element of the group.
+func (d Decaf) Generator() *Elt { return &Elt{*d.c.pull(Curve{}.Generator())} }
+
+// Order returns a scalar with the order of the group.
+func (d Decaf) Order() Scalar { return order }
+
+// Add calculates c=a+b, where + is the group operation.
+func (d Decaf) Add(c, a, b *Elt) { q := a.p; q.Add(&b.p); c.p = q }
+
+// Double calculates c=a+a, where + is the group operation.
+func (d Decaf) Double(c, a *Elt) { c.p = a.p; c.p.Double() }
+
+// Neg calculates c=-a, where - is the inverse of the group operation.
+func (d Decaf) Neg(c, a *Elt) { c.p = a.p; c.p.cneg(1) }
+
+// Mul calculates c=n*a, where * is scalar multiplication on the group.
+func (d Decaf) Mul(c *Elt, n *Scalar, a *Elt) { d.c.ScalarMult(&c.p, n, &a.p) }
+
+// MulGen calculates c=n*g, where * is scalar multiplication on the group,
+// and g is the generator of the group.
+func (d Decaf) MulGen(c *Elt, n *Scalar) { d.c.ScalarBaseMult(&c.p, n) }
+
+// IsIdentity returns True if e is the identity of the group.
+func (e *Elt) IsIdentity() bool { return fp.IsZero(&e.p.x) }
+
+// IsEqual returns True if e=a, where = is an equivalence relation.
+func (e *Elt) IsEqual(a *Elt) bool {
+	l, r := &fp.Elt{}, &fp.Elt{}
+	fp.Mul(l, &e.p.x, &a.p.y)
+	fp.Mul(r, &a.p.x, &e.p.y)
+	fp.Sub(l, l, r)
+	return fp.IsZero(l)
+}
+
+// UnmarshalBinary if succeeds returns a Decaf element by decoding the first
+// DecafEncodingSize bytes of data.
+func (e *Elt) UnmarshalBinary(data []byte) error {
+	if len(data) < DecafEncodingSize {
+		return errInvalidDecoding
+	}
+
+	s := &fp.Elt{}
+	copy(s[:], data[:DecafEncodingSize])
+	isPositiveS := fp.Parity(s) == 0
+	p := fp.P()
+	isLessThanP := isLessThan(s[:], p[:])
+
+	s2, den, num := &fp.Elt{}, &fp.Elt{}, &fp.Elt{}
+	isr, altx := &fp.Elt{}, &fp.Elt{}
+	t0, t1 := &fp.Elt{}, &fp.Elt{}
+	x, y := &fp.Elt{}, &fp.Elt{}
+	one := fp.One()
+	fp.Sqr(s2, s)                       // s2  = s^2
```

Comment:
Matching the paper notation would help verify the correctness of this code.

Reply:
Now matches [Hamburg spec](https://mailarchive.ietf.org/arch/msg/cfrg/S4YUTt_5eD4kwYbDuhEK0tXT1aM).